File: adapters/typescript/check.mjs
import fs from 'node:fs';
import path from 'node:path';
import { pathToFileURL } from 'node:url';

const clientsRoot = process.env.OPTO_SYNC_CLIENTS_DIR;
if (!clientsRoot) {
  throw new Error('OPTO_SYNC_CLIENTS_DIR is required');
}

const modulePath = path.join(
  clientsRoot,
  'clients',
  'ts',
  'dist',
  'esm',
  'schema',
  'ingest.js',
);
const { parseEnvelope } = await import(pathToFileURL(modulePath));
const { compareHlc, formatHlc, parseHlc } = await import(pathToFileURL(path.join(
  clientsRoot,
  'clients',
  'ts',
  'dist',
  'esm',
  'clock.js',
)));
const decisions = {};

for (const fixturePath of process.argv.slice(2)) {
  const key = `${path.basename(path.dirname(fixturePath))}/${path.basename(fixturePath)}`;
  try {
    await parseEnvelope(fs.readFileSync(fixturePath, 'utf8'));
    decisions[key] = true;
  } catch {
    decisions[key] = false;
  }
}

const formatted = formatHlc({ millis: 1721822400000, counter: 255, nodeId: '9f3a2b' });
const hlc = {
  formatted,
  parsed: parseHlc(formatted),
  compared: compareHlc(formatted, '1721822400001-0000-9f3a2b'),
};

process.stdout.write(`${JSON.stringify({ runtime: 'typescript', decisions, hlc })}\n`);

File: adapters/dart/check.dart
import 'dart:convert';
import 'dart:io';

import 'package:opto_sync_client/schema.dart' as schema;
import 'package:opto_sync_client/opto_sync_client.dart' as client;

void main(List<String> arguments) {
  final decisions = <String, bool>{};
  for (final fixturePath in arguments) {
    final file = File(fixturePath);
    final category = file.parent.uri.pathSegments
        .where((part) => part.isNotEmpty)
        .last;
    final key = '$category/${file.uri.pathSegments.last}';
    try {
      schema.parseEnvelope(file.readAsStringSync());
      decisions[key] = true;
    } catch (_) {
      decisions[key] = false;
    }
  }
  const parts = client.HlcParts(
    millis: 1721822400000,
    counter: 255,
    nodeId: '9f3a2b',
  );
  final formatted = client.formatHlc(parts);
  final parsed = client.parseHlc(formatted)!;
  final hlc = <String, Object>{
    'formatted': formatted,
    'parsed': <String, Object>{
      'millis': parsed.millis,
      'counter': parsed.counter,
      'nodeId': parsed.nodeId,
    },
    'compared': client.compareHlc(
      formatted,
      '1721822400001-0000-9f3a2b',
    ).sign,
  };
  stdout.writeln(
    jsonEncode({'runtime': 'dart', 'decisions': decisions, 'hlc': hlc}),
  );
}
